password/src/Data/Password/Validate.hs

```diff
@@ -218,18 +233,21 @@ isValidPasswordPolicy PasswordPolicy{..} =
     -- the character set contain at least 1 uppercase letter)
     requiredCharsetValid :: Bool
     requiredCharsetValid =
-      let charsets = charSet
-          hasUpperCase = hasChar uppercaseChars Uppercase charsets
-          hasLowerCase = hasChar lowercaseChars Lowercase charsets
-          hasSpecial = hasChar specialChars Special charsets
-          hasDigits = hasChar digitChars Digit charsets
-      in and [hasUpperCase, hasLowerCase, hasSpecial, hasDigits]
-    hasChar :: Maybe Int -> CharacterCategory -> Text -> Bool
-    hasChar Nothing _ _ = True
-    hasChar (Just _) category sets =
-      let predicate = case category of
-            Uppercase -> isAsciiUpper
-            Lowercase -> isAsciiLower 
-            Special   -> (\char -> char `elem` specialLetters)
-            Digit     -> isDigit
-      in T.any predicate sets
+      let predicates = accumulatePredicates
+            [ (uppercaseChars, isAsciiUpper)
+            , (lowercaseChars, isAsciiLower)
+            , (specialChars, (\c -> c `elem` specialCharacterSet))
+            , (digitChars, isDigit)
+            ]
+      in any (\c -> matchAny predicates c && (getCharSetPredicate charSetPredicate) c) [' ' .. '~']
```

Comment:
Made improvements to existing function (previous implementation was going over the `[Char]` multiple times which made it inefficient).

Reply:
I think this breaks if someone wants to only allow characters that are outside of the ASCII range.
I don't think the `charSetPredicate` should be `&&`-ed here. Just testing the accumulatedPredicates should suffice.
And I'd rather use the `defaultCharacterSet` list here instead of `[' ' .. '~']`

(Maybe also add a unit test that all the characters in `defaultCharacterSet` satisfy the `defaultCharSetPredicate`?)

Reply:
After taking at look at it, I think the whole logic is completely wrong. What I was trying to do here is to check that if `PasswordPolicy`  defines that it requires certain characters, then `CharSetPredicate` should return true on at least one of the characters that is required.

For instance, if `PasswordPolicy` states that the password requires at least one uppercase letter, then `CharSetPredicate` should return True on at least one uppercase letter.